hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestDeleteWithSlowFollower.java

```diff
@@ -0,0 +1,288 @@
+    client = OzoneClientFactory.getClient(conf);
+    objectStore = client.getObjectStore();
+    xceiverClientManager = new XceiverClientManager(conf);
+    volumeName = "testcontainerstatemachinefailures";
+    bucketName = volumeName;
+    objectStore.createVolume(volumeName);
+    objectStore.getVolume(volumeName).createBucket(bucketName);
+  }
+
+  /**
+   * Shutdown MiniDFSCluster.
+   */
+  @AfterClass
+  public static void shutdown() {
+    if (cluster != null) {
+      cluster.shutdown();
+    }
+  }
+
+  /**
+   * The test simulates a slow follower by first writing key thereby creating a
+   * a container on 3 dns of the cluster. Then, a dn is shutdown and a close
+   * container cmd gets issued so that in the leader and the alive follower,
+   * container gets closed. And then, key is deleted and
+   * the node is started up again so that it
+   * rejoins the ring and starts applying the transaction from where it left
+   * by fetching the entries from the leader. Until and unless this follower
+   * catches up and its replica gets closed,
+   * the data is not deleted from any of the nodes which have the
+   * closed replica.
+   */
+  @Test
+  public void testDeleteKeyWithSlowFollower() throws Exception {
+
+    OzoneOutputStream key =
+        objectStore.getVolume(volumeName).getBucket(bucketName)
+            .createKey("ratis", 0, ReplicationType.RATIS,
+                ReplicationFactor.THREE, new HashMap<>());
+    byte[] testData = "ratis".getBytes();
+    // First write and flush creates a container in the datanode
+    key.write(testData);
+    key.flush();
+
+    KeyOutputStream groupOutputStream = (KeyOutputStream) key.getOutputStream();
+    List<OmKeyLocationInfo> locationInfoList =
+        groupOutputStream.getLocationInfoList();
+    Assert.assertEquals(1, locationInfoList.size());
+    OmKeyLocationInfo omKeyLocationInfo = locationInfoList.get(0);
+    long containerID = omKeyLocationInfo.getContainerID();
+    // A container is created on the datanode. Now figure out a follower node to
+    // kill/slow down.
+    HddsDatanodeService follower = null;
+    HddsDatanodeService leader = null;
+
+    List<Pipeline> pipelineList =
+        cluster.getStorageContainerManager().getPipelineManager()
+            .getPipelines(HddsProtos.ReplicationType.RATIS,
+                HddsProtos.ReplicationFactor.THREE);
+    Assert.assertTrue(pipelineList.size() == 1);
+    Pipeline pipeline = pipelineList.get(0);
+    for (HddsDatanodeService dn : cluster.getHddsDatanodes()) {
+      if (ContainerTestHelper.isRatisFollower(dn, pipeline)) {
+        follower = dn;
+      } else if (ContainerTestHelper.isRatisLeader(dn, pipeline)) {
+        leader = dn;
+      }
+    }
+    Assert.assertNotNull(follower);
+    Assert.assertNotNull(leader);
+    //shutdown the slow follower
+    cluster.shutdownHddsDatanode(follower.getDatanodeDetails());
+    key.write(testData);
+    key.close();
+    XceiverClientSpi xceiverClient =
```

Comment:
Can we please add a comment here that the container will be closed? Similarly a comment regarding deleting the key just so that we can separate the various sections of the test.

Reply:
addressed in the latest patch..